clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleZonalServerGroupCachingAgent.java

```diff
@@ -0,0 +1,906 @@
+import com.netflix.spinnaker.clouddriver.google.provider.GoogleInfrastructureProvider;
+import com.netflix.spinnaker.clouddriver.google.security.GoogleNamedAccountCredentials;
+import com.netflix.spinnaker.clouddriver.names.NamerRegistry;
+import com.netflix.spinnaker.moniker.Moniker;
+import com.netflix.spinnaker.moniker.Namer;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public final class GoogleZonalServerGroupCachingAgent
+    implements CachingAgent, OnDemandAgent, AccountAware {
+
+  private static final ImmutableSet<AgentDataType> DATA_TYPES =
+      ImmutableSet.of(
+          AUTHORITATIVE.forType(SERVER_GROUPS.getNs()),
+          AUTHORITATIVE.forType(APPLICATIONS.getNs()),
+          INFORMATIVE.forType(CLUSTERS.getNs()),
+          INFORMATIVE.forType(LOAD_BALANCERS.getNs()));
+
+  private static final String ON_DEMAND_TYPE =
+      String.join(":", GoogleCloudProvider.getID(), OnDemandType.ServerGroup.name());
+
+  private static final Splitter COMMA = Splitter.on(',').omitEmptyStrings().trimResults();
+  private static final MapSplitter IMAGE_DESCRIPTION_SPLITTER =
+      Splitter.on(',').withKeyValueSeparator(": ");
+
+  private final GoogleNamedAccountCredentials credentials;
+  private final GoogleComputeApiFactory computeApiFactory;
+  private final String region;
+  private final OnDemandMetricsSupport onDemandMetricsSupport;
+  private final Namer<GoogleLabeledResource> naming;
+  private final ObjectMapper objectMapper;
+
+  public GoogleZonalServerGroupCachingAgent(
+      GoogleNamedAccountCredentials credentials,
+      GoogleComputeApiFactory computeApiFactory,
+      Registry registry,
+      String region,
+      ObjectMapper objectMapper) {
+    this.credentials = credentials;
+    this.computeApiFactory = computeApiFactory;
+    this.region = region;
+    this.onDemandMetricsSupport = new OnDemandMetricsSupport(registry, this, ON_DEMAND_TYPE);
+    this.objectMapper = objectMapper;
+    this.naming =
+        NamerRegistry.lookup()
+            .withProvider(GoogleCloudProvider.getID())
+            .withAccount(credentials.getName())
+            .withResource(GoogleLabeledResource.class);
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+
+    try {
+      CacheResultBuilder cacheResultBuilder = new CacheResultBuilder();
+      cacheResultBuilder.setStartTime(System.currentTimeMillis());
+
+      List<GoogleServerGroup> serverGroups = getServerGroups(providerCache);
+
+      // If an entry in ON_DEMAND was generated _after_ we started our caching run, add it to the
```

Comment:
Huh, yeah, that definitely seems like a crazy way of doing things. I can think of other adjectives too! But thanks for the backstory. I updated the comment; let me know if it appears inaccurate or incomplete:

```
// We don't evict things unless they've been processed because Orca expects to see these keys
// when calling pendingOnDemandRequests, and then knows the on-demand cache request has
// finished when they disappear. If they go away before Orca has a chance to see them, Orca
// decides the on-demand cache request was ignored completely and sends it again. So we have
// to leave them around a while so Orca can observe them. Is this weird? Yes. Yes, it is.
```

Reply:
Only slight comment is that orca doesn't know they've succeded "when they disappear" but rather when it sees they have a `processedCount` of >0, but other than that looks great!

Reply:
Ah, great, thanks! Updated the comment (again).